### DISCUSSION TITLE
Help to avoid wasm-ld: error: --shared-memory is disallowed by sfnt.c.o because ...

### DISCUSSION THREAD

**Discussion by cecilios** [emscripten-core/emscripten#16666] [Q&A] [answered]
I'm trying to build JavaScript bindins for the Lomse library (https://github.com/lenmus/lomse).

I built the library (static) with `emcmake cmake` using the following flags :
```
-sUSE_ZLIB=1 -sUSE_LIBPNG=1 -sUSE_FREETYPE=1 -sUSE_PTHREADS=1 -pthread
```
and I don't link to other third party libraries, appart from the zlib, libpng, freetype and pthreads provided by emscripten.

Then, I copied the generated liblomse.a and the library *.h files to the project for building the JavaScript bindigns, and tried to build with:
```
emcc -I../lomse/include --bind -sUSE_ZLIB=1 -sUSE_LIBPNG=1 -sUSE_FREETYPE=1 -sUSE_PTHREADS=1 -pthread -Wl,--shared-memory ../src/bindings.cpp ../lomse/liblomse.a -o lomse-js.js
```

But after many changes and attempts to build I always get this error:
```
wasm-ld: error: --shared-memory is disallowed by sfnt.c.o because it was not compiled with 'atomics' or 'bulk-memory' features.
```

The last attempt was to build all (the library sources and the bindings) in a single command, instead of building first the library:
```
emcc -I./ -I../include -I../src/agg/include -I../packages -I../packages/minizip \
-I../packages/pugixml -I../packages/utfcpp -I../packages/utfcpp/utf8 \
--bind -sUSE_ZLIB=1 -sUSE_LIBPNG=1 -sUSE_FREETYPE=1 -sUSE_PTHREADS=1 -pthread -Wl,--shared-memory \
../src/document/lomse_command.cpp \
../src/document/lomse_document.cpp \
../src/document/lomse_document_cursor.cpp \
../src/document/lomse_document_iterator.cpp \
../src/document/lomse_id_assigner.cpp \
    .... many lines with source files follows
../src/sound/lomse_score_player.cpp \
../packages/pugixml/pugixml.cpp \
../../../lomse-js/src/bindings.cpp -o lomse-js.js

```

But I still get the error:
```
wasm-ld: error: --shared-memory is disallowed by sfnt.c.o because it was not compiled with 'atomics' or 'bulk-memory' features.
```

Any help with this will be appreciated. I don't know what else to do or to change. Thank you.

**Comment by cecilios** [comment, marked as answer]
Solved. I don't know if this is the right approach but I solved this by not using threads and removing flags `-sUSE_PTHREADS=1 -pthread -Wl,--shared-memory`

**Comment by sbc100** [comment]
The other solution would be to ensure that `sfnt.c.o` is build with `-pthread` or `-sUSE_PTHREAD`
